ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AbstractRetrieverTool.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.engine.tools;
+
+import static org.opensearch.ml.common.utils.StringUtils.gson;
+
+import java.io.IOException;
+import java.security.AccessController;
+import java.security.PrivilegedExceptionAction;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.Client;
+import org.opensearch.common.xcontent.LoggingDeprecationHandler;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.ml.common.spi.tools.AbstractTool;
+import org.opensearch.ml.common.spi.tools.Tool;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+/**
+ * Abstract tool supports search paradigms in neural-search plugin.
+ */
+@Log4j2
+@Getter
+@Setter
+public abstract class AbstractRetrieverTool extends AbstractTool {
```

Comment:
This tool moved to skills repo https://github.com/opensearch-project/skills/blob/main/src/main/java/org/opensearch/agent/tools/AbstractRetrieverTool.java.

Reply:
It is confusing that tools are split between ml-commons and skills repo, it might better to move all tools to skills repo even if there are no dependencies to other plugins.

Reply:
I think keeping the common tools in ml-commons could make integ test of agent framework easier. Otherwise we have another circular dependency between ml-commons and skills repo. For common tools like agent tool and ml model tool, they are very lightweight. 

The logic is very like the way how we organize OpenSearch plugins: we have some common plugins in OpenSearch core ([code link](https://github.com/opensearch-project/OpenSearch/tree/main/plugins)) , but we also have separate repo for individual/heavy plugins.

Reply:
The plan is moving all tools depending on other plugins to the new skills repo. The new repo is mainly to solve the circular dependency between ml-commons and other plugins which depend on ml-commons. For example , SQL plugin depends on ml-commons client now, if we add PPL tool to ml-commons, then ml-commons will also depend on SQL plugin. That will be circular dependency for integ test. We discussed with infra team about other options like using custom build workflow to build ml-commons client, then build SQL, then build ml-commons plugin. But that's not easy from infra team and hard to maintain. That's why we build the new skills repo , so we can use current way to build ml-commons first, then the ml-commons client will be ready. Then we can build SQL plugin. When both ml-commons and SQL ready, we can build the new skills repo.

Reply:
I agree on skills repo being the owner of tools and it will help simplify tool management. However, maintaining some subset of tools in ml-commons just for integ tests doesn't seem a valid usecase. If required integ tests can create dummy tools to bypass circular dependency between ml-commons and skills package.